analytics.js

```diff
@@ -182,6 +185,50 @@ class Analytics {
     }
   }
 
+  /**
+   * A function to validate and return Residency server input
+   * @returns string/undefined
+   */
+  getResidencyServerInput() {
+    const region = get(this.options, 'residencyServer');
+    return typeof region === 'string' && RESIDENCY_SERVERS.includes(region.toUpperCase())
+      ? region
+      : undefined;
+  }
+
+  /**
+   * A function to return the dataPlaneUrl provided in load API(if available) or the default one
+   * @returns string
+   */
+  checkUrlFromCode() {
+    if (this.serverUrl && this.isValidServerUrl(this.serverUrl)) {
+      return this.serverUrl;
+    }
+    return DEFAULT_DATAPLANE_URL;
+  }
+
+  /**
+   * A function to determine the dataPlaneUrl for sending cloud mode calls
+   * @param {Object} dataPlaneUrls An object containing dataPlaneUrl for different region
+   * @returns string
+   */
+  resolveDataPlaneUrl(dataPlaneUrls = {}) {
+    // Check if dataPlaneUrls object is present in source config
+    if (Object.keys(dataPlaneUrls).length) {
+      const inputRegion = this.getResidencyServerInput() || '';
+      const dataPlaneUrl =
+        dataPlaneUrls[inputRegion.toUpperCase()] || dataPlaneUrls[inputRegion.toLowerCase()];
+      if (dataPlaneUrl) {
+        return dataPlaneUrl;
+      }
+      if (dataPlaneUrls[DEFAULT_REGION]) {
+        return dataPlaneUrls[DEFAULT_REGION];
```

Comment:
We should check if the user provided URLs are also valid.

Reply:
Do you mean URLs provided in source config?

Reply:
Yes.

Reply:
Addressed